### ISSUE TITLE
extend EffectBus for world-process intents (soil, water, climate)

### ISSUE THREAD

**Issue by biosynthart** [hellolifeforms/lila#63]
## Context

The engine currently runs world-level processes as inline methods on `EcosystemEngine`:
- `_evaporate_soil(dt)` — O(grid²) moisture walk every tick
- `_replenish_water_sources(dt)` — water level + soil footprint update every tick
- `_apply_voxel_effects(entity, dt)` — entity→soil nutrient/moisture drain per entity
- `apply_rain(intensity)` — global event handler (moisture/nutrient boost, entity hydration)

These are **physics/chemistry processes** that conceptually run at different natural frequencies than the entity tick:

| Process | Natural frequency | Current frequency |
|---|---|---|
| Soil evaporation | ~2 Hz (slow thermal process) | 10 Hz (every tick) |
| Water source replenishment | ~1 Hz (groundwater is slow) | 10 Hz (every tick) |
| Nutrient diffusion | ~5 Hz (chemical transport) | 10 Hz (per entity, no diffusion between cells) |
| Entity→soil drain/deposit | 10 Hz (coupled to entity state) | 10 Hz ✓ correct |

The EffectBus already exists and handles entity-level effects (`StateVarDelta`, `SetTarget`, `ClearTarget`, `Spawn`, `Remove`). Extending it for world-process intents is the bridge between today's inline methods and tomorrow's partitioned, multi-frequency processes.

### 1. New effect types (world-process intents)

```python
# effects.py — new world-level effect dataclasses

class SoilEvaporation(Effect):
    """Intent: evaporate soil moisture based on climate conditions."""
    tick: int
    temperature: float
    humidity: float
    rain_suppressed: bool  # True during RAIN_SUPPRESSION_TICKS

class WaterReplenish(Effect):
    """Intent: update water source levels and soil footprints."""
    tick: int
    sources: list[dict]  # water source dicts (mutated in-place by handler)

class SoilDrain(Effect):
    """Intent: entity drains nutrients/moisture from overlapping grid cells."""
    tick: int
    entity_id: str
    position: list[float]
    layer: str  # "nutrients" or "moisture"
    amount: float  # negative delta to apply

class SoilDeposit(Effect):
    """Intent: deposit organic matter into overlapping grid cells."""
    tick: int
    entity_id: str
    position: list[float]
    layer: str  # "organic_matter"
    amount: float

class RainEvent(Effect):
    """Intent: apply rainfall across the world (soil + water sources + entities)."""
    tick: int
    intensity: float
```

### 2. Process handler protocol

```python
# effects.py — new protocol for process handlers

class WorldProcessHandler(Protocol):
    """Handles a world-process effect type at its own frequency."""

    # Which effect types this handler consumes
    handles: tuple[type[Effect], ...]

    # How often to run (in ticks). 1 = every tick, 5 = every 5th tick.
    period: int = 1

    def resolve(self, effect: Effect, context: WorldProcessContext) -> None:
        """Apply the world-process effect.

        Args:
            effect: The intent to process (SoilEvaporation, WaterReplenish, etc.)
            context: Read-only access to voxels, entities, climate, rate multipliers.
        """
        ...
```

### 3. WorldProcessContext

```python
# effects.py — read-only context for world-process handlers

class WorldProcessContext:
    tick: int
    voxel_grid: VoxelManager
    biome: BiomeConfig
    climate: dict[str, float]
    entities: dict[str, dict]  # full entity registry (read-only)
    water_sources: list[dict]
    rate_multipliers: dict[str, float]
```

### 4. EffectBus extension

```python
class EffectBus:
    # Existing entity-level registries...
    self._entity_handlers = {...}

    # New world-process handler registry
    self._world_handlers: list[WorldProcessHandler] = []

    def register_world_handler(self, handler: WorldProcessHandler) -> None:
        """Register a world-process handler with its frequency."""
        self._world_handlers.append(handler)

    def apply_world_batch(
        self,
        effects: list[Effect],
        tick: int,
        context: WorldProcessContext,
    ) -> None:
        """Dispatch world-process effects to registered handlers.

        Each handler only runs when tick % period == 0, enabling
        multi-frequency process execution without engine-side bookkeeping.
        """
        for effect in effects:
            for handler in self._world_handlers:
                if type(effect) not in handler.handles:
                    continue
                if tick % handler.period != 0:
                    continue  # Skip — not this handler's turn
                handler.resolve(effect, context)
```

### 5. Concrete handlers (extracted from engine inline methods)

```python
# world_processes.py — new module

class SoilEvaporationHandler(WorldProcessHandler):
    """Handles soil moisture evaporation at ~2 Hz."""
    handles = (SoilEvaporation,)
    period = 5  # runs every 5th tick → 2 Hz at 10 Hz tick rate

    def resolve(self, effect: SoilEvaporation, ctx: WorldProcessContext) -> None:
        if effect.rain_suppressed:
            return
        temp = effect.temperature
        humidity = effect.humidity
        evap_rate = (SOIL_EVAP_BASE_RATE * (temp / SOIL_EVAP_TEMP_SCALE)
                     * (1.0 - humidity * SOIL_EVAP_HUMIDITY_FACTOR) * ctx.rate_multipliers["thirst"])
        dims = ctx.voxel_grid.dimensions
        for x in range(dims[0]):
            for z in range(dims[2]):
                current = ctx.voxel_grid.get("moisture", x, 0, z)
                if current > SOIL_MOISTURE_FLOOR:
                    ctx.voxel_grid.set("moisture", x, 0, z,
                                       max(SOIL_MOISTURE_FLOOR, current - evap_rate))


class WaterReplenishHandler(WorldProcessHandler):
    """Handles water source replenishment at ~1 Hz."""
    handles = (WaterReplenish,)
    period = 10  # runs every 10th tick → 1 Hz at 10 Hz tick rate

    def resolve(self, effect: WaterReplenish, ctx: WorldProcessContext) -> None:
        for source in effect.sources:
            evap_loss = WATER_EVAPORATION_RATE * ctx.rate_multipliers["thirst"]
            replenish = WATER_REPLENISH_RATE * ctx.rate_multipliers["water_replenishment"]
            source["water_level"] = max(0.0, min(1.0,
                source["water_level"] - evap_loss + replenish))
            # ... soil footprint update (same logic as current _replenish_water_sources)
```

### 6. Engine step() after extraction

```python
def step(self, dt: float = 0.1) -> dict[str, Any]:
    self.tick += 1
    # ... phases 1-3 unchanged (flow, movement, interactions, guards)

    # Phase 4: Voxel effects — entity impact on soil (now effect-based)
    voxel_effects = []
    for entity in list(self.entities.values()):
        if is_alive(entity):
            effects = self._build_voxel_effects(entity, dt)  # emits SoilDrain/SoilDeposit
            voxel_effects.extend(effects)
    world_ctx = WorldProcessContext(
        tick=self.tick,
        voxel_grid=self.voxels,
        biome=self.biome,
        climate=self.climate,
        entities=self.entities,
        water_sources=self.water_sources,
        rate_multipliers={...},
    )
    self.effect_bus.apply_world_batch(voxel_effects, self.tick, world_ctx)

    # Phase 5: Water & soil — world-level processes (now effect-based)
    world_effects = [
        SoilEvaporation(tick=self.tick,
                       temperature=self.climate.get("temperature", 20.0),
                       humidity=self.climate.get("humidity", 0.5),
                       rain_suppressed=self._rain_ticks_remaining > 0),
        WaterReplenish(tick=self.tick, sources=self.water_sources),
    ]
    self.effect_bus.apply_world_batch(world_effects, self.tick, world_ctx)

    # ... phases 6-7 unchanged (motor, spawn/kill)
```

## Benefits

1. **Multi-frequency execution** — handlers run at their natural cadence via `period`, no engine-side bookkeeping needed
2. **Testability** — each handler is unit-testable with a mock `WorldProcessContext` (no full engine required)
3. **Extensibility** — new processes (nutrient diffusion, temperature gradients, wind transport) are just new handlers + effect types
4. **Multi-resolution grid ready** — `SoilDrain`/`SoilDeposit` effects carry entity position + amount; the handler queries which cells overlap. When VoxelManager becomes adaptive, only the handler changes (cell overlap query), not the engine or effect type
5. **Cleaner engine** — removes ~80 lines of inline world-process code from `EcosystemEngine`

## Migration path

1. Add new effect types to `effects.py` (no behavior change)
2. Add `WorldProcessHandler` protocol + `apply_world_batch()` to EffectBus
3. Create `world_processes.py` with handlers extracted from engine inline methods
4. Register handlers in engine `__init__`
5. Replace inline Phase 4/5 code in `step()` with effect emission + bus dispatch
6. Delete `_evaporate_soil`, `_replenish_water_sources`, `_apply_voxel_effects` from engine
7. Tune handler periods for desired frequencies (start matching current behavior: period=1)
8. Gradually increase periods to natural frequencies once validated

## Scope

- New effect types in `effects.py`: ~60 lines
- `WorldProcessHandler` protocol + `WorldProcessContext`: ~40 lines
- EffectBus extension (`apply_world_batch`, handler registry): ~30 lines
- New file `world_processes.py` (extracted handlers): ~120 lines
- Engine changes: remove ~80 inline lines, add ~25 effect emission lines
- Total net change: ~75 lines added, ~80 removed ≈ neutral

## Related issues

- #64: multi-resolution voxel grid interface (SoilDrain/SoilDeposit effects are the bridge)
- #62: hot path cleanup findings (superseded by this issue + #64)
